Caching http client

craig-day opened the pull request:
Motivation:

This adds a web client interceptor to support caching HTTP responses based on a number of properties. The goal is to help clients improve performance, and to help reduce the load clients place on upstream services responding to the requests. 

**TODO:**

- [ ] I'd like to write an abstract document explaining the approach taken here, maybe even with some nice diagrams. I'm not sure the best place to put something like this.
- [x] Fix the response handling when refreshing stale responses. Right now, it caches the new response in place of the old one, but a server may send back a 304 Not Modified response with no body, which means we would cache the new empty body, instead of refreshing the existing cached body.
- [x] Do we want to support `stale-while-revalidate` 
- [x] Do we want to support `stale-while-error`?
- ~Do we prune stale responses? This is only necessary if we don't implement the stale handling above~
- [x] Do we want to fully implement support for `Cache-Control: no-cache`? Meaning we store the response, but validate with the server each time. This is only beneficial if we add support for `stale-while-revalidate` or `stale-if-error` support.
- [ ] I'd love to build in instrumentation support, so any pointers as to what/where to look for pointers or examples would be much appreciated.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/cache/CacheAdapter.java:
```diff
@@ -0,0 +1,54 @@
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public interface CacheAdapter {
```
it should be renamed to simply Cache or WebCache

craig-day replied:
I renamed it to `CacheStore` to be consistent with `CookieStore`. `Cache` feels overly generic, but I can rename it to that if you would prefer.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CachedHttpRequestImpl.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CachedHttpRequestImpl extends HttpRequestImpl<Buffer> {
```
can we instead use full delegation here, instead of inheriting HttpRequestImpl ?

craig-day replied:
I changed this to `implements HttpRequest<Buffer>` and then delegate to a `HttpRequestImpl` buffer. Is that what you meant by "full delegation"?

vietj replied:
yes

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/WebClientCacheAware.java:
```diff
@@ -0,0 +1,105 @@
+/**
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class WebClientCacheAware extends WebClientBase implements WebClientCache {
```
rather CachingWebClient

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CachedHttpRequestImpl.java:
```diff
@@ -0,0 +1,99 @@
+    WebClientCacheOptions options, CacheAdapter cacheAdapter) {
+    super(client, method, serverAddress, ssl, port, host, uri, codec, options);
+    this.options = options;
+    this.cacheManager = new CacheManager(cacheAdapter, options);
```
why do we need a new cache manager for every request ?

craig-day replied:
We don't, this was just an early design artifact. I've removed the creation of the manager up to the client level, so we will get one manger per client instead of per request now.

vietj commented:
CacheStore looks great and coherent

On Mon, Apr 26, 2021 at 7:11 PM Craig Day ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> vertx-web-client/src/main/java/io/vertx/ext/web/client/cache/CacheAdapter.java
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620490516>:
>
> > + *  You may elect to redistribute this code under either of these licenses.
> + */
> +package io.vertx.ext.web.client.cache;
> +
> +import io.vertx.core.AsyncResult;
> +import io.vertx.core.Future;
> +import io.vertx.core.Handler;
> +import io.vertx.ext.web.client.impl.cache.CacheKey;
> +import io.vertx.ext.web.client.impl.cache.CachedHttpResponse;
> +
> +/**
> + * An API to store and retrieve HTTP responses.
> + *
> + * @author <a ***@***.***">Craig Day</a>
> + */
> +public interface CacheAdapter {
>
> I renamed it to CacheStore to be consistent with CookieStore. Cache feels
> overly generic, but I can rename it to that if you would prefer.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620490516>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCS72JLWAS4YUEUS7Q3TKWNDRANCNFSM43PIBVGA>
> .
>

vietj commented:
yes totally, this way the cache store is decoupled from the implementation

On Mon, Apr 26, 2021 at 7:11 PM Craig Day ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CachedHttpRequestImpl.java
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620490972>:
>
> > +import io.vertx.core.Future;
> +import io.vertx.core.Handler;
> +import io.vertx.core.buffer.Buffer;
> +import io.vertx.core.http.HttpMethod;
> +import io.vertx.core.net.SocketAddress;
> +import io.vertx.ext.web.client.HttpRequest;
> +import io.vertx.ext.web.client.HttpResponse;
> +import io.vertx.ext.web.client.WebClientCacheOptions;
> +import io.vertx.ext.web.client.cache.CacheAdapter;
> +import io.vertx.ext.web.client.impl.cache.CacheManager;
> +import io.vertx.ext.web.codec.BodyCodec;
> +
> +/**
> + * @author <a ***@***.***">Craig Day</a>
> + */
> +public class CachedHttpRequestImpl extends HttpRequestImpl<Buffer> {
>
> I changed this to implements HttpRequest<Buffer> and then delegate to a
> HttpRequestImpl buffer. Is that what you meant by "full delegation"?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620490972>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCRIFYSRO56ZTY5PB5TTKWNF5ANCNFSM43PIBVGA>
> .
>

vietj commented:
thanks, this looked weird and I was wondering what I was missing

On Mon, Apr 26, 2021 at 7:12 PM Craig Day ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In
> vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CachedHttpRequestImpl.java
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620491534>:
>
> > + */
> +public class CachedHttpRequestImpl extends HttpRequestImpl<Buffer> {
> +
> +  private final CacheManager cacheManager;
> +  private final WebClientCacheOptions options;
> +
> +  static CachedHttpRequestImpl wrap(HttpRequest<Buffer> request, CacheAdapter cacheAdapter) {
> +    return new CachedHttpRequestImpl((HttpRequestImpl<Buffer>) request, cacheAdapter);
> +  }
> +
> +  CachedHttpRequestImpl(WebClientInternal client, HttpMethod method, SocketAddress serverAddress,
> +    Boolean ssl, Integer port, String host, String uri, BodyCodec<Buffer> codec,
> +    WebClientCacheOptions options, CacheAdapter cacheAdapter) {
> +    super(client, method, serverAddress, ssl, port, host, uri, codec, options);
> +    this.options = options;
> +    this.cacheManager = new CacheManager(cacheAdapter, options);
>
> We don't, this was just an early design artifact. I've removed the
> creation of the manager up to the client level, so we will get one manger
> per client instead of per request now.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1932#discussion_r620491534>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCWRCNJLHRHGID6XRP3TKWNITANCNFSM43PIBVGA>
> .
>

vietj commented:
can you rebase the PR, it looks like there is a conflict with `WebClientTest` class

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java:
```diff
@@ -0,0 +1,131 @@
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public interface CachingWebClient extends WebClient {
```
I don't think there is a real interest to have a specific `CachingWebClient` interface that extends `WebClient` as it does not have specific operations.

craig-day replied:
Does this mean it should not extend `WebClient`, and if so what should it do instead, or are you suggesting that we add the cache features into the core WebClient itself and just make it configurable?

vietj replied:
yes the interface should not extend the WebClient but the implementation should still extend WebClient of course. It is just that we do not need a new functionnal interface for this.

vietj replied:
so the implementation that wraps the actual WebClient still implements WebClient

craig-day replied:
Ok. So how would someone get the caching implementation? Should it be via config, or a few some new functions on the `WebClient` interface?

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClientOptions.java:
```diff
@@ -0,0 +1,212 @@
+/**
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CachingWebClientOptions extends WebClientOptions {
```
properties should use get/set pattern, in addition CachingWebClientOptions should be annotated with `@DataObject`

craig-day replied:
For boolean properties, I followed the pattern from `WebClientOptions` where instead of `getBooleanProperty` it was `isBooleanProperty`. Is it OK to leave the `is*` pattern here, or should I do `getBooleanProperty` for this class?

vietj replied:
I think *is* is fine but we can check the generated converter

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControl.java:
```diff
@@ -0,0 +1,154 @@
+
+  private final Set<String> directives;
+  private final Map<String, Long> timeDirectives;
+  private final Date expires;
```
can we use the new immutable date / time API instead of java.util.Date ?

craig-day replied:
yeah, I think so. I did this because the netty class that can parse/encode http date headers uses java.util.Date, but I think it will be straightforward enough to convert to/from them.

vietj replied:
thanks

craig-day replied:
Updated this to use `Instant` and `Duration`, please un-resolve this comment if something else is desired.

craig-day commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java:
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2021 Red Hat, Inc.
```
@vietj is this the correct form of header? The examples reference either putting my name in or a generic "The authors" statement.

vietj replied:
you can use the header that gives copyright to the "Contributors to the Eclipse Foundation", look at `WebClientSession`

vietj commented:
this is looking good, as far as I can see there is no real cache store implemented yet

craig-day commented:
Yeah, I wasn't sure if we want to ship one out of the box, and if we do what it should use. Thoughts on either just an in memory `ConcurrentHashMap` or maybe leveraging the shared data store?

craig-day commented:
Pushed up an implementation and used it as a default if one isn't given. I realized that I'm handling some scenarios incorrectly here by trying to match a request header to a response header, but that isn't how it works in every case. It should match a past request header to a current request header. The reference talks about embedding those aspects into the key, but that makes querying hard. I think I can address this by changing `CachedHttpResponse` to store the headers of the request that triggered the cache to compare with future requests. I'm working on that now.

vietj commented:
can you rebase on latest master to remove CI errors ?

craig-day commented:
It looks like I'm on latest, and it seems like the errors still exist there. Am I missing something?

vietj commented:
I will have a look

vietj commented:
I reran CI and it looks like there are codegen build errors

vietj commented:
ping @craig-day can you fix the codegen errors in the build ?

vietj commented:
thanks, I reran CI as windows did fail.

I'll do a new review pass.

craig-day commented:
Thanks @vietj . I marked this PR as ready for review now since I think I've got all the features added that I wanted. I'd still love to add some instrumentation SPI type thing, but I'm not sure if there are currently any patterns around that. I'd also be happy to write up conceptual docs (to talk about how it works and any caveats that might exist, like around Vary responses) for this if desired, but I don't know where those would go.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControlDirectives.java:
```diff
@@ -0,0 +1,31 @@
+
+import io.vertx.codegen.annotations.VertxGen;
+
+@VertxGen
```
no need for `@VertxGen` here

vietj commented:
I think we need to move the interception logic to an interceptor so the caching web-client can be used along with the session `WebClientSession`

craig-day commented:
@vietj I'm not sure we can use an interceptor with an async cache store. I think I tried this when first starting, but the next loop proceeded to the next phase before my interceptor finished its async execution. I'll look again since I haven't looked at it since Vertx 3.

vietj commented:
yes please, as interceptor is internal API we could modify it to make
it work I think


On Wed, Jun 16, 2021 at 5:49 PM Craig Day ***@***.***> wrote:
>
> @vietj I'm not sure we can use an interceptor with an async cache store. I think I tried this when first starting, but the next loop proceeded to the next phase before my interceptor finished its async execution. I'll look again since I haven't looked at it since Vertx 3.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

vietj commented:
@craig-day the interceptor has been much improved since vertx 3 and the interceptor as control over the next phase to execute

craig-day commented:
Sounds great. I'll definitely work on switching to an interceptor over the next few days. I probably won't have time to work on it until the weekend unfortunately.

vietj commented:
sure thing, we are on the right path

On Thu, Jun 17, 2021 at 7:33 PM Craig Day ***@***.***> wrote:
>
> Sounds great. I'll definitely work on switching to an interceptor over the next few days. I probably won't have time to work on it until the weekend unfortunately.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

craig-day commented:
Okay, @vietj I made a lot of progress using the interceptor. Most it it is straightforward but there are two big things: 

- I added a new `ClientPhase`, `SEND_REQUEST`. I did this because I needed a distinction between the `HttpResponse` being built and being sent to the handler. The cache wants to store the `HttpResponse` so it needed to be after dispatch, but then when I want to return a cached result calling dispatch would put the interceptor into an infinite loop. 
- I don't know how to handle the "revalidation request" pattern. For example, if a cached value is stale. In this case, we want to send the request to the server, and if the server replies with a 304 we just reply with the cached value and mark it as valid. However, using the interceptor the only way I can send the request is to advance the phase, which means I don't reply to the initial request with the cached value. I think I can make this work by adjusting receive response and if we receive a 304, look for a cached value and assume this request was revalidating that. I'll play with that later, but I'm not sure how to handle the `stale-if-error or the `stale-while-revalidate` case.

vietj commented:
@craig-day for the revalidation can you record this in the `attrs` map so later you know whether it was a revalidation ?

craig-day commented:
@vietj thanks for that pointer, it helped a ton. I've removed my extra `ClientPhase` and just set an attribute on the context for that. I was also able to do almost all of the revalidation work with attrs as well. The last thing I'm not sure about is the `stale-while-revalidate` support. Currently, I call `context.clientRequest().send()` from the interceptor. This seems to trigger a new context and new request to get executed and I just don't need to block on it, then my interceptor can carry on with what it's doing. It seems OK, but I'm not sure if there are potential bad side effect from this.

vietj commented:
I'm glad you could make progress this way

you can operate directly on the request if you need to

I'll review that in the PR anyway

On Wed, Jun 23, 2021 at 9:28 PM Craig Day ***@***.***> wrote:
>
> @vietj thanks for that pointer, it helped a ton. I've removed my extra ClientPhase and just set an attribute on the context for that. I was also able to do almost all of the revalidation work with attrs as well. The last thing I'm not sure about is the stale-while-revalidate support. Currently, I call context.clientRequest().send() from the interceptor. This seems to trigger a new context and new request to get executed and I just don't need to block on it, then my interceptor can carry on with what it's doing. It seems OK, but I'm not sure if there are potential bad side effect from this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

craig-day commented:
Sounds good! I think I'm happy with where this is and don't plan on iterating further until PR review. Thanks for all the help.

vietj commented:
@craig-day great, this PR should be merged after 4.1.1 release when we branch for 4.1 and master becomes 4.2

vietj commented:
@craig-day now Vert.x master is 4.2 and we can finish review this PR and merge it for Vert.x 4.2.0 release

vietj commented:
what is the current status of the PR from your perspective @craig-day ?

craig-day commented:
@vietj I'm happy with the state of this PR, so I think it's ready for full review. Thanks!

vietj commented:
alright I'll start the review tomorrow

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/CachingWebClientImpl.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CachingWebClientImpl extends WebClientBase {
```
I think we do not need to extend WebClientBase anymore here

vietj replied:
reminder

craig-day replied:
Removed this class entirely.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/WebClientBase.java:
```diff
@@ -259,6 +259,9 @@ public WebClientBase(HttpClient client, WebClientOptions options) {
   public WebClientInternal addInterceptor(Handler<HttpContext<?>> interceptor) {
+    // If a web client is constructed using another client, interceptors could get added twice.
+    // This is an attempt to ensure that doesn't happen by keeping the most recent one added.
+    interceptors.removeIf(i -> i.getClass() == interceptor.getClass());
```
instead we should throw an exception such as IllegalStateException and let the client code know

vietj replied:
well it could work if client for instance double wrap the same client, better be aware of the error rather than silently try to make some magic

craig-day replied:
Done, throws an exception now.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControlDirectives.java:
```diff
@@ -0,0 +1,28 @@
+ */
+package io.vertx.ext.web.client.impl.cache;
+
+public interface CacheControlDirectives {
```
I think we should use an enum here that wraps the string value

craig-day replied:
Done.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java:
```diff
@@ -0,0 +1,228 @@
+  private void handleSendRequest(HttpContext<Buffer> context) {
+    HttpRequestImpl<Buffer> requestImpl = (HttpRequestImpl<Buffer>) context.request();
+    if (!HttpMethod.GET.equals(requestImpl.method())) {
+      context.next();
```
we can group this statement and the next one

```java
Vary variation;
if (HttpMethod.GET != requestImpl.method() || (variation = selectVariation(requestImpl)) == null) {
  context.next()
  return;
}
```

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java:
```diff
@@ -0,0 +1,228 @@
+
+  private void handleDispatchResponse(HttpContext<Buffer> context) {
+    Boolean isCacheDispatch = context.get(IS_CACHE_DISPATCH);
+    if (isCacheDispatch != null && isCacheDispatch) {
```
instead `isCacheDispatch == Boolean.TRUE`

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java:
```diff
@@ -0,0 +1,228 @@
+    } else {
+      // We're storing a new response in cache
+      processResponse(context, null).onComplete(ar -> {
+        // TODO: is this right? if we don't need the data then use a callback, not a future
```
would processResponse fail ?

vietj replied:
if we don't care about the response failure then it is fine

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java:
```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  private Future<HttpResponse<Buffer>> respondFromCache(HttpContext<Buffer> context, CachedHttpResponse response) {
```
I think here we could distinguish the absence of response from a failure with `Future<Optional<HttpResponse<Buffer>>>`

craig-day replied:
Didn't need the future at all, was just using it to `compose` another future, so I switched to `map` instead and just returns an optional now.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheKey.java:
```diff
@@ -0,0 +1,85 @@
+
+  @Override
+  public String toString() {
+    String rawKey = super.toString() + "|" + variations;
```
we should avoid concatenating then computing the digest and instead compute the digest of the three strings:

```java
digest.update(super.toString().getBytes(StandardCharsets.UTF_8));
digest.update("|".getBytes(StandardCharsets.UTF_8));
digest.update(variations.getBytes(StandardCharsets.UTF_8));
byte[] hashed = digest.digest();
```

I think we should even try to do something more optimized here with a faster algorithm here and avoid using a toString()

vietj replied:
is that actually called often or use for comparison ?

vietj replied:
i.e do we care about performance here ?

craig-day replied:
That's partly an implementation detail of the CacheStore. If you're using a remote store like redis or memcached, then you'll want a nice concise string to use as the key. In that case, this will get called for each get/set. I generally prefer opaque keys for cases like this. I could probably use `Objects.hash` instead of a digest library if we're concerned about performance here.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CachedHttpResponse.java:
```diff
@@ -0,0 +1,125 @@
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public class CachedHttpResponse implements Serializable {
```
can we avoid using `Serializable` and let the actual cache deal with serialization when needed

craig-day replied:
Hm, why do we want to avoid it? How would a cache store know which attributes to serialize to ensure that the cached response they return will work?

vietj replied:
cache store should have access to the various field of the CachedHttpResponse and be responsible for the serialization.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java:
```diff
@@ -0,0 +1,154 @@
+ *
+ * @author <a href="mailto:craigday3@gmail.com">Craig Day</a>
+ */
+public interface CachingWebClient {
```
should be annotated with `@VertxGen`

craig-day commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheInterceptor.java:
```diff
@@ -0,0 +1,228 @@
+      // Request was successful, attempt to cache response
+      return cacheResponse(request, response).map(response);
+    } else if (cachedResponse != null && cachedResponse.useStaleIfError()) {
+      // TODO: we should check that this is an error too
```
we should check that `context.response()` was an error response before checking the `cachedResponse.useStaleIfError()` flag. If the response didn't result in an error, we should just use that instead of falling back to the cached response.

craig-day commented:
I'm running into some codegen problems now. @vietj can you help me with the codegen stuff?

vietj commented:
@craig-day sure, add comment about the issues you have and I'll help you

craig-day commented:
I'm not sure how to generate the JSON converter for `CachingWebClientOptions`. I added `@VertxGen` to the `CachingWebClient` interface, and `@DataObject` to `CachingWebClientOptions`. The first error I got said I needed a JSON constructor and then a toJSON function, so I added those and copied the format from `WebClientOptions`, but I'm not sure how to generate the `CachingWebClientOptionsConverter` that they use. Can you either tell me how to (or link some docs) run the codegen, or can we just remove `@DataObject`?

vietj commented:
I will create a copy of your branch and make codegen pass for you

vietj commented:
@craig-day after a quick look at the code, I think the only missing thing is to add `@GenIgnore(GenIgnore.PERMITTED_TYPE)` on `CachingWebClient` methods declaring the `CacheStore` type

vietj commented:
can you rebase the branch on lastest master ?

craig-day commented:
Thanks for the pointer @vietj . Squashed some commits, rebased on latest, and fix the codegen with that GenIgnore annotation

craig-day commented:
@vietj anything else I need to do here? Just want to make sure you aren't waiting on me for anything.

vietj commented:
@craig-day I will have a look again at the PR soon, no worries

vietj commented:
Overall I would like to understand and discuss how caching applies to HTTP users and the interactions between the caching interceptor and `WebClientSession`.

Currently as I see it, either the application will wrap a shared WebClient with a cache or do the same with `WebClientSession`. But I would like to hear the full story on your side about this :-)

craig-day commented:
My initial thought is that they should mix together reasonably well as I'm not sure caching cares much about the session headers. However, I don't know a lot of details around the session and all of the things it does. I could see there being a concern about caching a response longer than a session might be valid, but due to the nature of it being a cached result, we aren't really at risk of over-exposing information I think. Not sure if that makes sense. 

The easiest option is to make them exclusive and either say use the cache or the session-aware client, but I don't know if that's necessary. Looking at the session interceptor, it looks like it stores cookies and then applies them to future requests. I don't think this conflicts with the caching interceptor explicitly. 

The session interceptor operates on `CREATE_REQUEST` and `DISPATCH_RESPONSE` phases, whereas the cache interceptor operates on `SEND_REQUEST` and `DISPATCH_RESPONSE`. Both interceptors only store information to an external store during dispatch and that will continue to work as desired.

The biggest area I'm not sure about is what happens when the session interceptor sets headers during `CREATE_REQUEST` and then the cache interceptor dispatches a cached response during `SEND_REQUEST`. Since we aren't sending the request to the upstream now, does this even matter? Maybe this means when the session interceptor handles the dispatch, it won't be able to store the data it expects. Do you have thoughts on this? 

To summarize, I don't think having both interceptors will matter here, but I might be missing something that the session interceptor does.

vietj commented:
@craig-day I'm also referring to CacheControl private and public directives, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching

I think we need to honor that and if a WebClient is configured with a cache then the session should be used to store private caching instead of the shared cache, or find something similar.

vietj commented:
ping @craig-day

craig-day commented:
Hey @vietj, sorry missed the last notification. I think I see what you're saying. Right now, the client can be configured to enable/disable caching private things, but is the concern that it isn't good enough? 

Are you thinking that rather than having configuration for private cache, we should see if we're in a session aware client? I suppose if we can make the clients aware of each other, then we can store private responses with the session information, otherwise we only store public responses in the shared cache. Does that seem more desirable? If so, should we and how do we make the interceptors aware of each other?

vietj commented:
@craig-day you hit the nail, private mode should use the web client session, when it is not available then we need to think of what should be the behavior. 

Can you check whether it is easy currently to use the session when available for private caching ?

craig-day commented:
I'll look into this today and tomorrow. Finally understanding what the best behavior is :D

vietj commented:
thanks @craig-day I'm eager to see the progress of the final steps!

craig-day commented:
After an initial look I have some questions/thoughts. I'm not sure how we would best make the interceptors aware of each other. I think we can put something into the context to help with that, but that might introduce an order dependency of which interceptor runs the create_request state first. 

A bigger question I have is where do we actually store the cached response? If we have a session interceptor, it has a `CookieStore`, but that isn't flexible enough to store a HTTP response. Similarly, the cache store for the cache interceptor doesn't know about the cookies, but I think I could embed the cookies into the cache key by hashing them or something.

All of that to say that I don't think it's an "easy" thing to just add, because neither of the stores can handle both the cookies and the http response for caching. I'm not sure which API it makes more sense to change, or maybe I'm going in the wrong direction with this.

vietj commented:
I think we need to have `WebClientSession` and `HttpContext` have a `CacheStore` field and have the session set it on the context.

craig-day commented:
> have the session set it on the context

I'm not sure what that means. I started working on a change to add a `CacheStore` field to the context, but the rest I got lost on. What role does `WebClientSession` play here? I think I'm confused on if you think the session should control the caching, or if the `CacheInterceptor` should just be made aware of the session information. 

What I started doing was: 
- Added a `CacheStore` field to the `HttpContext`
- Added a `PREPARE_REQUEST` hook to the `CacheInterceptor` and if `context.cacheStore()` is null, then set the cache store on the context
- Change the `SessionAwareInterceptor` to add the encoded cookie string to the context attrs during `CREATE_REQUEST`
- Update the `CacheInterceptor` to check the attrs for a cookie during `SEND_REQUEST` and if it is present then include that in the cache key

Does any of this match the direction you were thinking? 

On an unrelated note, I think the cache implementation needs to change in the `Vary` case. Right now, there is a local map that maps a variation to a cache key, but that is on the instance of the `CacheInterceptor`. I think that will effectively mean that varying caches will be less effective because the information will only be scoped to one instance of the client. Does that seem like a concern, or do you think clients are lived long enough for that to be OK?

craig-day commented:
@vietj yes that makes a lot of sense! I'm working on that now. Thanks for the clear thoughts.

craig-day commented:
Oh, sorry. I pushed up a first attempt. Yeah, I think it could use a review.

vietj commented:
thanks I will soon

On Sat, Aug 28, 2021 at 1:34 AM Craig Day ***@***.***> wrote:

> Oh, sorry. I pushed up a first attempt. Yeah, I think it could use a
> review.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-web/pull/1932#issuecomment-907525926>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABXDCQHGVFOJ4ETEIWPEK3T7AOJ5ANCNFSM43PIBVGA>
> .
>

vietj commented:
I'll have a look tomorrow

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/LocalCacheStore.java:
```diff
@@ -0,0 +1,33 @@
+
+public class LocalCacheStore implements CacheStore {
+
+  private final Map<CacheKey, CachedHttpResponse> localMap = new ConcurrentHashMap<>();
```
a regular hashmap is fine here

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/HttpContext.java:
```diff
@@ -56,6 +56,7 @@
   private final Handler<AsyncResult<HttpResponse<T>>> handler;
   private final HttpClientImpl client;
   private final List<Handler<HttpContext<?>>> interceptors;
+  private final CacheStore privateCacheStore;
```
I think we should not make this final and have a mutator so we don't need to expose the state and constructor to the session web client implementation

vietj commented:
I did a quick review of the last commit, and it seems to do the right thing. I added a couple of comments and I think after that we need documentation and examples before doing a final review

craig-day commented:
Sounds good. I'll work on that over the next few days. When you say "need documentation and examples", where would I write that?

vietj commented:
you should look at the src/main/asciidoc/index.adoc in the
vertx-web-client directory, examples are coded in Java in
src/main/java/examples

thanks!

On Fri, Sep 3, 2021 at 6:40 PM Craig Day ***@***.***> wrote:
>
> Sounds good. I'll work on that over the next few days. When you say "need documentation and examples", where would I write that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

craig-day commented:
@vietj sounds good! I was working on it over the weekend and working more today. Should hopefully have a commit pushed up today.

craig-day commented:
Ok. @vietj I pushed up some simple examples and docs. Let me know what you think.

pmlopes commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/CachingWebClient.java:
```diff
@@ -0,0 +1,162 @@
+   * @param vertx the vertx instance
+   * @return the creatd web client
+   */
+  static WebClient create(Vertx vertx) {
```
We should agree on a factory default. Session aware and OAuth2 aware clients always wrap a preconfigured base web client in a method named `create(...)` and we don't have a `wrap(...)` variant. 

If we decide that we should simplify it for users to have a `create()` to create a base client and wrap, then we should also apply the same style to the other wrappers we already have.

IMHO, we are favouring composition over inheritance, so my personal choice (which is totally open for discussion) would be to remove the current `create()` factories and rename the `wrap(...)` factories to `create(...)` which would make it "consistent" with the other clients

pmlopes replied:
/cc @vietj for the comment above

vietj replied:
I agree with you @pmlopes

craig-day replied:
@pmlopes and @vietj , just to be clear here do we want `CachingWebClient` to be just like `WebClientSession`? Meaning it should `extends WebClient`, or just renaming the static interface methods to match name/signature? Related, should the caching client implementation then `extends WebClientBase`, or can it just keep the two methods it has where it just adds the interceptor?

pmlopes replied:
@craig-day, no. Your PR is ok, what I recommend is a doing 2 small changes:

1. remove the all `create` methods
2. rename the remaining `wrap` to `create`

For example it should end up looking like this:

```java
package io.vertx.ext.web.client;

...

@VertxGen
public interface CachingWebClient {

  /**
   * Create a {@link HttpClient} with a web client and default options.
   *
   * @param httpClient the http client to wrap
   * @param cacheStore the cache adapter
   * @return the created web client
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static WebClient create(HttpClient httpClient, CacheStore cacheStore) {
    CachingWebClientOptions options = new CachingWebClientOptions();
    return wrap(httpClient, cacheStore, options);
  }

  /**
   * Create a {@link HttpClient} with a web client and custom options.
   *
   * @param httpClient the http client to wrap
   * @param cacheStore the cache adapter
   * @param options    the caching web client options
   * @return the created web client
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static WebClient create(HttpClient httpClient, CacheStore cacheStore, CachingWebClientOptions options) {
    return CachingWebClientImpl.wrap(httpClient, cacheStore, options);
  }
}
```

As this is the same API style already used by the WebClientOAuth2: https://github.com/vert-x3/vertx-web/blob/master/vertx-web-client/src/main/java/io/vertx/ext/web/client/WebClientOAuth2.java

Finally, we may want to agree on the interface name, so it also means we should rename the new `WebClientOAuth2` interface to `OAuth2WebClient` to keep the naming consistent:

* WebClient (base)
* CachingWebClient (this)
* OAuth2WebClient (the other wrapper added only in master atm)

craig-day replied:
Thanks for the details @pmlopes. I cleaned up the in interface to 4 `create` methods. I think it's nice to have a method where you don't have to give a store and a default in-memory store is used. If we don't want this, I can slim the interface down to just the two methods you suggested. /cc @vietj

pmlopes reviewed (changes requested):
Please see my note on the factory methods. If we agree on an outcome, functionality wise I think the PR is good.

vietj commented on vertx-web-client/src/main/asciidoc/index.adoc:
```diff
@@ -464,6 +464,58 @@ The client will follow at most `16` requests redirections, it can be changed in
+----
+{@link examples.CachingWebClientExamples#createWithSession(io.vertx.core.Vertx)}
+----
+
```
I think we need to indicate that when private and public caching are enabled, private caching takes precedence over public caching

craig-day replied:
What do you mean here? We should only use the private store when we see `Cache-Control: private`. Are you referring to the code comment saying about seeing `Cache-Control: public, private` and saying we pick private? I think that's technically in invalid case, but I can add a mention of it to the docs if you would still like me to.

vietj commented on vertx-web-client/src/main/java/examples/CachingWebClientExamples.java:
```diff
@@ -0,0 +1,63 @@
+
+  public void createWithOptions(Vertx vertx) {
+    Set<HttpMethod> cachedMethods = new HashSet<>(2);
+    cachedMethods.add(HttpMethod.GET);
```
use instead `addCachedMethod` on options directly

vietj commented on vertx-web-client/src/main/java/examples/CachingWebClientExamples.java:
```diff
@@ -0,0 +1,63 @@
+    cachedMethods.add(HttpMethod.HEAD);
+
+    Set<Integer> cachedStatusCodes = new HashSet<>(1);
+    cachedStatusCodes.add(200);
```
use instead `addCachedStatusCode` on options directly

craig-day replied:
I actually wanted to demonstrate reducing the status codes, so I changed this to use `removeCachedStatusCode` directly instead of creating a new set.

vietj replied:
makes sense

vietj commented on vertx-web-client/src/main/java/examples/CachingWebClientExamples.java:
```diff
@@ -0,0 +1,63 @@
+
+  public void createWithCustomStore(Vertx vertx) {
+    WebClient client = WebClient.create(vertx);
+    CacheStore store = new NoOpCacheStore();
```
Add a comment that say "or any store you like"

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControl.java:
```diff
@@ -0,0 +1,176 @@
+      return Collections.emptySet();
+    } else {
+      Set<CharSequence> variations = new HashSet<>();
+      for (String variation : vary.split(",")) {
```
we should avoid using `split()` for performance reason, instead we should split this using string indexOf(char)

craig-day replied:
Updated, let me know if you're happier with the new splitting.

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControl.java:
```diff
@@ -0,0 +1,176 @@
+  }
+
+  public boolean isCacheable() {
+    if (directives.contains(CacheControlDirective.NO_STORE)) {
```
use single statement with `|`

vietj commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/cache/CacheControl.java:
```diff
@@ -0,0 +1,176 @@
+    }
+
+    return maxAge > 0;
+  }
```
I think we can use instead `return !(directives.contains(CacheControlDirective.NO_STORE) | "*".equals(vary)) && maxAge > 0`

craig-day replied:
I'll defer to you on this, but out of curiosity is this for a performance reason? It feels harder to read to me. Related, do we not want a logical or here so the statement can short circuit?

vietj replied:
it does not matter much indeed, more my taste but I'll leave like that

vietj reviewed (changes requested):
Couple of minor changes

craig-day commented:
@vietj I pushed a round of updates and had some questions.

pmlopes commented:
Thank ylou @craig-day for your contribution! :+1:

vietj commented:
thanks a lot @craig-day for the involvement and the time spent here, that makes a nice feature for the community for sure :-)
